keps/1397-evictions-in-background/README.md

```diff
@@ -0,0 +1,1122 @@
+  Other components may create an evacuation request as well.
+  The descheduler needs to take this into account.
+  E.g. by creating a snapshot (per cycle, per strategy) to avoid inconsistencies.
+  An evacuation request may exist without a corresponding pod
+  before the evacuation controller garbage collects the request.
+- The descheduler needs to keep a cache of evacution API requests (CRs)
+  and remove itself from finalizers if the request is no longer needed.
+  E.g. `evacuation.coordination.k8s.io/instigator_descheduler.sigs.k8s.io` as a finalizer.
+- Read the eviction status from the [evacuation CR status](https://github.com/kubernetes/enhancements/pull/4565).
+
+#### Code changes
+
+The code responsible for invoking the eviction API is located
+under `sigs.k8s.io/descheduler/pkg/descheduler/evictions` under private `evictPod` function.
+The function can be generalized into a generic interface:
+
+```go
+type PodEvacuator interface {
+    EvacutePod(ctx context.Context, pod *v1.Pod) error
+}
+```
+
+to provide the [descheduling framework](https://github.com/kubernetes-sigs/descheduler/pull/1372)
+a generic way of evicting/evacuating pods.
+The new interface can be used for implementing various testing scenarios.
+Or turned into a plugin if needed.
+
+#### Metrics
+
+A new metric `pod_evacuations` for counting the number of evacuated pods is introduced.
+The descheduler will not observe whether a pod was ultimately evicted.
+Only provide a summary about how many pods were requested to be evacuated.
+If needed, another metric `pod_evacuations_in_progress` for counting
+the number of evacutions that are reconciled can be introduced.
+
+### Open Questions [optional]
+
+This is where to call out areas of the design that require closure before deciding
+to implement the design.
+
+* The evacution API proposal mentions the evacuation instigator should remove
+  its intent when the evacuation is no longer necessary.
+  Should each descheduling cycle reset all the eviction requests
+  or wait until some of the eviction requests were completed
+  to free a "bucket" for new evictions?
+  **Answer**: The first implementation will not reset any evacuation request.
+  The descheduler will account for existing requests and update the internal counters accordingally.
+  In the future a mechanism for deleting too old evacuation requests can be introduced.
+  I.e. based on a new `--max-evacuation-request-ttl` option.
+* The evacution API proposal mentions more than a one entity can request an eviction.
+  Should the descheduler take these into account as well?
+  What if another entity decides to evict a pod that is of a low priority
+  from the descheduler's point of view?
+  **Answer**: The first implementation will consider all pods with an existing
+  evacuation request as "already getting evacuated". Independent of the pod priorities.
+* With evacution requests a plugin does not evict a pod right away.
```

Comment:
Is it possible to move to just using the Evacuations? Would it be simpler to reason about by the plugins?

Reply:
Can you more elaborate on what you mean? In my mind evacuations = evacuation requests.

Reply:
I forgot what I meant originally.

Can we pass the Evacuation objects through the plugins then and have them decide on their state?

Reply:
The eviction/evacuation is exposed/expected to be exposed through a generic interface. All the knowledge about how eviction/evacuation processes are implemented is currently expected to be exposed through filters/sorters. So no plugin implementation depends on a particular eviction/evacuation implementation. Some community adopters would like to turn the eviction/evacuation handle into a plugin as well.